File: clojure/gradle-build-logic/java-plugins/src/main/clojure/software/onepiece/gradle/pluginshowto/MyJavaBasePlugin.clj
(ns software.onepiece.gradle.pluginshowto.MyJavaBasePlugin
    (:import
      (com.diffplug.gradle.spotless SpotlessPlugin SpotlessExtension)
      (org.gradle.api Project Plugin, Action)
      (org.gradle.api.plugins JavaPlugin JavaPluginExtension)
      (org.gradle.api.tasks.testing Test)
      (org.gradle.jvm.toolchain JavaLanguageVersion)
    )
    (:gen-class
      :name       software.onepiece.gradle.pluginshowto.MyJavaBasePlugin
      :implements [org.gradle.api.Plugin])
    )

(defn -apply [this ^Project project]
  (do
    (.apply (.getPlugins project) JavaPlugin)
    (.apply (.getPlugins project) SpotlessPlugin)

    ; Configure Java compilation
    (.set (.getLanguageVersion (.getToolchain (.getByType (.getExtensions project) JavaPluginExtension))) (JavaLanguageVersion/of 17))

    ; Configure JUnit5 as test framework
    (.named (.getTasks project) "test" (reify Action (execute [this test] (do
      (.useJUnitPlatform test)
      (.showStackTraces (.getTestLogging test) true)
    ))))
    (.add (.getDependencies project) "implementation" "org.junit.jupiter:junit-jupiter:5.7.2")

    ; Configure a community plugin - example Spotless
    (.format (.getByType (.getExtensions project) SpotlessExtension) "buildFiles" (reify Action (execute [this format] (do
      (.target format (into-array Object ["build.gradle.kts"]))
      (.trimTrailingWhitespace format)
      (.endWithNewline format)
    ))))
  )
)

File: java/gradle-build-logic/java-plugins/src/main/java/software/onepiece/gradle/pluginshowto/MyJavaBasePlugin.java
package software.onepiece.gradle.pluginshowto;

import com.diffplug.gradle.spotless.SpotlessExtension;
import com.diffplug.gradle.spotless.SpotlessPlugin;
import org.gradle.api.NonNullApi;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.artifacts.dsl.DependencyHandler;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.api.plugins.JavaPluginExtension;
import org.gradle.api.tasks.testing.Test;
import org.gradle.jvm.toolchain.JavaLanguageVersion;

import static org.gradle.api.plugins.JavaPlugin.TEST_IMPLEMENTATION_CONFIGURATION_NAME;

@NonNullApi
abstract public class MyJavaBasePlugin implements Plugin<Project> {

    @Override
    public void apply(Project project) {
        project.getPlugins().apply(JavaPlugin.class);
        project.getPlugins().apply(SpotlessPlugin.class);

        // Configure Java compilation
        JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
        java.getToolchain().getLanguageVersion().set(JavaLanguageVersion.of(17));

        // Configure JUnit5 as test framework
        project.getTasks().named("test", Test.class, test -> {
            test.useJUnitPlatform();
            test.getTestLogging().setShowStackTraces(true);
        });
        project.getDependencies().add(TEST_IMPLEMENTATION_CONFIGURATION_NAME, "org.junit.jupiter:junit-jupiter:5.7.2");
    }
}
